File: app/src/main/java/com/xpf/p2p/jpush/LocalBroadcastManager.java
package com.xpf.p2p.jpush;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.Uri;
import android.os.Handler;
import android.os.Message;


import com.xpf.p2p.utils.LogUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Set;

/**
 * Created by efan on 2017/4/14.
 */

public final class LocalBroadcastManager {

    private static final String TAG = "LocalBroadcastManager";
    private static final boolean DEBUG = false;
    private final Context mAppContext;
    private final HashMap<BroadcastReceiver, ArrayList<IntentFilter>> mReceivers = new HashMap<BroadcastReceiver, ArrayList<IntentFilter>>();
    private final HashMap<String, ArrayList<ReceiverRecord>> mActions = new HashMap<String, ArrayList<ReceiverRecord>>();
    private final ArrayList<BroadcastRecord> mPendingBroadcasts = new ArrayList<BroadcastRecord>();
    static final int MSG_EXEC_PENDING_BROADCASTS = 1;
    private final Handler mHandler;
    private static final Object mLock = new Object();
    private static LocalBroadcastManager mInstance;

    public static LocalBroadcastManager getInstance(Context context) {
        Object var1 = mLock;
        synchronized (mLock) {
            if (mInstance == null) {
                mInstance = new LocalBroadcastManager(context.getApplicationContext());
            }

            return mInstance;
        }
    }

    private LocalBroadcastManager(Context context) {
        this.mAppContext = context;
        this.mHandler = new Handler(context.getMainLooper()) {
            @Override
            public void handleMessage(Message msg) {
                switch (msg.what) {
                    case 1:
                        LocalBroadcastManager.this.executePendingBroadcasts();
                        break;
                    default:
                        super.handleMessage(msg);
                }

            }
        };
    }

    public void registerReceiver(BroadcastReceiver receiver, IntentFilter filter) {
        HashMap var3 = this.mReceivers;
        synchronized (this.mReceivers) {
            LocalBroadcastManager.ReceiverRecord entry = new LocalBroadcastManager.ReceiverRecord(filter, receiver);
            ArrayList filters = (ArrayList) this.mReceivers.get(receiver);
            if (filters == null) {
                filters = new ArrayList(1);
                this.mReceivers.put(receiver, filters);
            }

            filters.add(filter);

            for (int i = 0; i < filter.countActions(); ++i) {
                String action = filter.getAction(i);
                ArrayList entries = (ArrayList) this.mActions.get(action);
                if (entries == null) {
                    entries = new ArrayList(1);
                    this.mActions.put(action, entries);
                }

                entries.add(entry);
            }

        }
    }

    public void unregisterReceiver(BroadcastReceiver receiver) {
        HashMap var2 = this.mReceivers;
        synchronized (this.mReceivers) {
            ArrayList filters = (ArrayList) this.mReceivers.remove(receiver);
            if (filters != null) {
                for (int i = 0; i < filters.size(); ++i) {
                    IntentFilter filter = (IntentFilter) filters.get(i);

                    for (int j = 0; j < filter.countActions(); ++j) {
                        String action = filter.getAction(j);
                        ArrayList receivers = (ArrayList) this.mActions.get(action);
                        if (receivers != null) {
                            for (int k = 0; k < receivers.size(); ++k) {
                                if (((LocalBroadcastManager.ReceiverRecord) receivers.get(k)).receiver == receiver) {
                                    receivers.remove(k);
                                    --k;
                                }
                            }

                            if (receivers.size() <= 0) {
                                this.mActions.remove(action);
                            }
                        }
                    }
                }

            }
        }
    }

    public boolean sendBroadcast(Intent intent) {
        HashMap var2 = this.mReceivers;
        synchronized (this.mReceivers) {
            String action = intent.getAction();
            String type = intent.resolveTypeIfNeeded(this.mAppContext.getContentResolver());
            Uri data = intent.getData();
            String scheme = intent.getScheme();
            Set categories = intent.getCategories();
            boolean debug = (intent.getFlags() & 8) != 0;
            if (debug) {
                LogUtils.v(TAG, "Resolving type " + type + " scheme " + scheme + " of intent " + intent);
            }

            ArrayList entries = (ArrayList) this.mActions.get(intent.getAction());
            if (entries != null) {
                if (debug) {
                    LogUtils.v(TAG, "Action list: " + entries);
                }

                ArrayList receivers = null;

                int i;
                for (i = 0; i < entries.size(); ++i) {
                    LocalBroadcastManager.ReceiverRecord receiver = (LocalBroadcastManager.ReceiverRecord) entries.get(i);
                    if (debug) {
                        LogUtils.v(TAG, "Matching against filter " + receiver.filter);
                    }

                    if (receiver.broadcasting) {
                        if (debug) {
                            LogUtils.v(TAG, "  Filter\'s target already added");
                        }
                    } else {
                        int match = receiver.filter.match(action, type, scheme, data, categories, "LocalBroadcastManager");
                        if (match >= 0) {
                            if (debug) {
                                LogUtils.v(TAG, "  Filter matched!  match=0x" + Integer.toHexString(match));
                            }

                            if (receivers == null) {
                                receivers = new ArrayList();
                            }

                            receivers.add(receiver);
                            receiver.broadcasting = true;
                        } else if (debug) {
                            String reason;
                            switch (match) {
                                case -4:
                                    reason = "category";
                                    break;
                                case -3:
                                    reason = "action";
                                    break;
                                case -2:
                                    reason = "data";
                                    break;
                                case -1:
                                    reason = "type";
                                    break;
                                default:
                                    reason = "unknown reason";
                            }

                            LogUtils.v(TAG, "Filter did not match: " + reason);
                        }
                    }
                }

                if (receivers != null) {
                    for (i = 0; i < receivers.size(); ++i) {
                        ((LocalBroadcastManager.ReceiverRecord) receivers.get(i)).broadcasting = false;
                    }

                    this.mPendingBroadcasts.add(new LocalBroadcastManager.BroadcastRecord(intent, receivers));
                    if (!this.mHandler.hasMessages(1)) {
                        this.mHandler.sendEmptyMessage(1);
                    }

                    return true;
                }
            }

            return false;
        }
    }

    public void sendBroadcastSync(Intent intent) {
        if (this.sendBroadcast(intent)) {
            this.executePendingBroadcasts();
        }
    }

    private void executePendingBroadcasts() {
        while (true) {
            LocalBroadcastManager.BroadcastRecord[] brs = null;
            HashMap i = this.mReceivers;
            synchronized (this.mReceivers) {
                int br = this.mPendingBroadcasts.size();
                if (br <= 0) {
                    return;
                }

                brs = new LocalBroadcastManager.BroadcastRecord[br];
                this.mPendingBroadcasts.toArray(brs);
                this.mPendingBroadcasts.clear();
            }

            for (int var6 = 0; var6 < brs.length; ++var6) {
                LocalBroadcastManager.BroadcastRecord var7 = brs[var6];

                for (int j = 0; j < var7.receivers.size(); ++j) {
                    ((LocalBroadcastManager.ReceiverRecord) var7.receivers.get(j)).receiver.onReceive(this.mAppContext, var7.intent);
                }
            }
        }
    }

    private static class BroadcastRecord {
        final Intent intent;
        final ArrayList<ReceiverRecord> receivers;

        BroadcastRecord(Intent _intent, ArrayList<ReceiverRecord> _receivers) {
            this.intent = _intent;
            this.receivers = _receivers;
        }
    }

    private static class ReceiverRecord {
        final IntentFilter filter;
        final BroadcastReceiver receiver;
        boolean broadcasting;

        ReceiverRecord(IntentFilter _filter, BroadcastReceiver _receiver) {
            this.filter = _filter;
            this.receiver = _receiver;
        }

        @Override
        public String toString() {
            StringBuilder builder = new StringBuilder(128);
            builder.append("Receiver{");
            builder.append(this.receiver);
            builder.append(" filter=");
            builder.append(this.filter);
            builder.append("}");
            return builder.toString();
        }
    }
}

File: app/src/main/java/com/xpf/p2p/jpush/LocalBroadcastManager.kt
package com.xpf.p2p.jpush

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Handler
import android.os.Message
import com.xpf.p2p.utils.LogUtils

/**
 * Created by efan on 2017/4/14.
 */
class LocalBroadcastManager private constructor(context: Context) {

    private val mAppContext: Context = context
    private val mReceivers = HashMap<BroadcastReceiver, ArrayList<IntentFilter>>()
    private val mActions = HashMap<String, ArrayList<ReceiverRecord>>()
    private val mPendingBroadcasts = ArrayList<BroadcastRecord>()
    private val mHandler: Handler

    init {
        mHandler = object : Handler(context.mainLooper) {
            override fun handleMessage(msg: Message) {
                when (msg.what) {
                    MSG_EXEC_PENDING_BROADCASTS -> executePendingBroadcasts()
                    else -> super.handleMessage(msg)
                }
            }
        }
    }

    fun registerReceiver(receiver: BroadcastReceiver, filter: IntentFilter) {
        synchronized(mReceivers) {
            val entry = ReceiverRecord(filter, receiver)
            var filters = mReceivers[receiver]
            if (filters == null) {
                filters = ArrayList(1)
                mReceivers[receiver] = filters
            }
            filters.add(filter)

            for (i in 0 until filter.countActions()) {
                val action = filter.getAction(i)
                var entries = mActions[action]
                if (entries == null) {
                    entries = ArrayList(1)
                    mActions[action] = entries
                }
                entries.add(entry)
            }
        }
    }

    fun unregisterReceiver(receiver: BroadcastReceiver) {
        synchronized(mReceivers) {
            val filters = mReceivers.remove(receiver) ?: return
            for (i in 0 until filters.size) {
                val filter = filters[i]
                for (j in 0 until filter.countActions()) {
                    val action = filter.getAction(j)
                    val receivers = mActions[action]
                    if (receivers != null) {
                        var k = 0
                        while (k < receivers.size) {
                            if (receivers[k].receiver === receiver) {
                                receivers.removeAt(k)
                                k--
                            }
                            k++
                        }
                        if (receivers.size <= 0) {
                            mActions.remove(action)
                        }
                    }
                }
            }
        }
    }

    fun sendBroadcast(intent: Intent): Boolean {
        synchronized(mReceivers) {
            val action = intent.action
            val type = intent.resolveTypeIfNeeded(mAppContext.contentResolver)
            val data = intent.data
            val scheme = intent.scheme
            val categories = intent.categories
            val debug = intent.flags and 8 != 0
            if (debug) {
                LogUtils.v(TAG, "Resolving type $type scheme $scheme of intent $intent")
            }

            val entries = mActions[intent.action]
            if (entries != null) {
                if (debug) {
                    LogUtils.v(TAG, "Action list: $entries")
                }

                var receivers: ArrayList<ReceiverRecord>? = null

                for (i in 0 until entries.size) {
                    val receiver = entries[i]
                    if (debug) {
                        LogUtils.v(TAG, "Matching against filter ${receiver.filter}")
                    }

                    if (receiver.broadcasting) {
                        if (debug) {
                            LogUtils.v(TAG, "  Filter's target already added")
                        }
                    } else {
                        val match = receiver.filter.match(action, type, scheme, data, categories, "LocalBroadcastManager")
                        if (match >= 0) {
                            if (debug) {
                                LogUtils.v(TAG, "  Filter matched!  match=0x${Integer.toHexString(match)}")
                            }
                            if (receivers == null) {
                                receivers = ArrayList()
                            }
                            receivers.add(receiver)
                            receiver.broadcasting = true
                        } else if (debug) {
                            val reason = when (match) {
                                -4 -> "category"
                                -3 -> "action"
                                -2 -> "data"
                                -1 -> "type"
                                else -> "unknown reason"
                            }
                            LogUtils.v(TAG, "Filter did not match: $reason")
                        }
                    }
                }

                if (receivers != null) {
                    for (i in 0 until receivers.size) {
                        receivers[i].broadcasting = false
                    }
                    mPendingBroadcasts.add(BroadcastRecord(intent, receivers))
                    if (!mHandler.hasMessages(MSG_EXEC_PENDING_BROADCASTS)) {
                        mHandler.sendEmptyMessage(MSG_EXEC_PENDING_BROADCASTS)
                    }
                    return true
                }
            }
            return false
        }
    }

    fun sendBroadcastSync(intent: Intent) {
        if (sendBroadcast(intent)) {
            executePendingBroadcasts()
        }
    }

    private fun executePendingBroadcasts() {
        while (true) {
            val brs: Array<BroadcastRecord>
            synchronized(mReceivers) {
                val n = mPendingBroadcasts.size
                if (n <= 0) return
                brs = mPendingBroadcasts.toTypedArray()
                mPendingBroadcasts.clear()
            }
            for (br in brs) {
                for (j in 0 until br.receivers.size) {
                    br.receivers[j].receiver.onReceive(mAppContext, br.intent)
                }
            }
        }
    }

    private class BroadcastRecord(val intent: Intent, val receivers: ArrayList<ReceiverRecord>)

    private class ReceiverRecord(val filter: IntentFilter, val receiver: BroadcastReceiver) {
        var broadcasting: Boolean = false

        override fun toString(): String {
            return "Receiver{$receiver filter=$filter}"
        }
    }

    companion object {
        private const val TAG = "LocalBroadcastManager"
        private const val DEBUG = false
        const val MSG_EXEC_PENDING_BROADCASTS = 1
        private val mLock = Any()
        private var mInstance: LocalBroadcastManager? = null

        @JvmStatic
        fun getInstance(context: Context): LocalBroadcastManager {
            synchronized(mLock) {
                if (mInstance == null) {
                    mInstance = LocalBroadcastManager(context.applicationContext)
                }
                return mInstance!!
            }
        }
    }
}
